File: _vendor/gluwa-usc-sdk/package/src/bin/encode-blocks.ts
import { mkdirSync, writeFileSync } from 'fs';
import { Block, WebSocketProvider, TransactionReceipt } from 'ethers';
import { abiEncode } from '../encoding/abi';
import { getTransactionWithRaw } from '../encoding';
import { bytesInHexString } from '../utils/hex';

// Maximum discovered size of ABI-encoded transaction data, in bytes.
// Derived from the largest observed successfully-encoded transactions on mainnet,
// which all topped out at exactly 449280 bytes of combined ABI-encoded data.
// Reference: gluwa/usc-testnet-bridge-examples#77 — decode-testing/top_abi_size_transactions.csv
//   block 25238768, tx 0x01ca130bf04e636d26ebdf0f6256a99894a6b474d4c016af74849c6a7572928d
//   block 25238750, tx 0x343b91c47944693ed1cdf3c979bd7722ed9284320ff6069bcfd46c109d9c4199
//   block 25238749, tx 0x5f60979ee18aba3f76122574e987f974fb1d7bacc372666f4b3f647236d54794
//   block 25238746, tx 0xf2641f3bd13a111169c007205b3d1e7188201df3ae041991d2e1e3745ed1fb2d
const MAX_ENCODED_SIZE = 449280;

/**
 * Gets all transaction receipts for a given block using regular Infura/compatible RPC format.
 * Uses hexadecimal block number format (e.g., "0x1a2b3c").
 *
 * @param rpc - The ethers JsonRpcApiProvider instance
 * @param block - The block to get receipts for
 * @returns Array of transaction receipts for the block
 */
async function getBlockReceipts(rpc: WebSocketProvider, block: Block): Promise<TransactionReceipt[]> {
  // Regular Infura/compatible RPC takes block number as hexadecimal string
  const finalBlockNumber = `0x${block.number.toString(16)}`;

  // Call the RPC method to get all receipts for the block
  const receiptsRaw: Array<any> = await rpc.send('eth_getBlockReceipts', [finalBlockNumber]);

  // Wrap raw receipts into proper TransactionReceipt objects
  const receipts = receiptsRaw.map((r) => {
    const receipt = rpc._wrapTransactionReceipt(r, rpc._network);
    return receipt;
  });

  return receipts;
}

// cost 80 or 160 credits depnding on arguments
async function encodeTransaction(
  provider: WebSocketProvider,
  txHash: string,
  receipt: TransactionReceipt | null,
): Promise<string> {
  // 80 credits
  const transaction = await getTransactionWithRaw(provider, txHash);

  if (receipt === null) {
    // 80 credits
    receipt = await provider.getTransactionReceipt(txHash);
  }
  const encodedData = abiEncode(transaction!, receipt!);
  return encodedData.abi;
}

async function encodeAndWriteToDisk(
  pathToStoreJson: string,
  provider: WebSocketProvider,
  blockNumber: number,
  txHash: string,
  receipt: TransactionReceipt | null,
) {
  const encodedData = await encodeTransaction(provider, txHash, receipt);

  const encodedSize = bytesInHexString(encodedData);
  if (encodedSize > MAX_ENCODED_SIZE) {
    throw new Error(
      `encoded data exceeds MAX_ENCODED_SIZE: blockNumber=${blockNumber} txHash=${txHash} encodedSize=${encodedSize} bytes (max=${MAX_ENCODED_SIZE})`,
    );
  }

  writeFileSync(`${pathToStoreJson}/${blockNumber}/${txHash}.txt`, encodedData + '\n', {
    flag: 'w',
  });
}

// https://docs.metamask.io/services/get-started/pricing/credit-cost#standard-ethereum-compliant-methods
// cost: (80 + 160 * <txns in block>)
async function blockHandler(
  prefix: string,
  provider: WebSocketProvider,
  blockNumber: number,
  pathToStoreJson: string,
): Promise<void> {
  console.log(`new ${prefix} --- ${blockNumber}`);

  mkdirSync(`${pathToStoreJson}/${blockNumber}`, { recursive: true });

  // cost: 80 credits
  const block = await provider.getBlock(blockNumber);
  const txHashes = block?.transactions || [];

  if (txHashes?.length >= 13) {
    // optimize RPC cost by fetching all receipts at once
    const receipts = await getBlockReceipts(provider, block!);

    await Promise.all(
      receipts.map(async (receipt) => {
        await encodeAndWriteToDisk(pathToStoreJson, provider, blockNumber, receipt.hash, receipt);
      }),
    );
  } else {
    // loop over each transaction and fetch its receipt via explicit RPC call
    await Promise.all(
      txHashes.map(async (txHash) => {
        await encodeAndWriteToDisk(pathToStoreJson, provider, blockNumber, txHash, null);
      }),
    );
  }

  console.log(`<<< done encoding ${txHashes?.length} transactions in block ${blockNumber}`);
}

async function encodeBlocks(rpcUrl: string, pathToStoreJson: string): Promise<void> {
  const start = Date.now();
  const timeoutMinutes = parseInt(process.env.TIMEOUT_MINUTES || '2');
  console.log(`=== starting with timeout ${timeoutMinutes} mins ...`);

  mkdirSync(pathToStoreJson, { recursive: true });

  const provider = new WebSocketProvider(rpcUrl);

  // Fire the event whenever the block changes.
  // We can also fire on 'safe' or 'finalized' blocks
  provider.on('block', async (blockNumber) => {
    try {
      // cost: (80 + 160 * <txns in block>)
      await blockHandler('block', provider, blockNumber, pathToStoreJson);
    } catch (err) {
      // A transient RPC error (e.g. `-32000 internal error` from the upstream
      // provider) should not kill the whole run. Log it and move on to the
      // next block — we'd rather skip one block than abort 60 minutes of work.
      console.error(`!!! skipping block ${blockNumber} due to error:`, err);
    }

    if (Math.floor((Date.now() - start) / 60000) >= timeoutMinutes) {
      console.log(`=== ${timeoutMinutes} mins timeout reached. exiting ...`);
      process.exit(0);
    }
  });
}

if (process.argv.length < 4) {
  console.error('node dist/bin/encode-blocks.js <ws://ethRpcUrl> <pathToStoreJson>');
  process.exit(1);
}

const rpcUrl = process.argv[2] || 'ws://127.0.0.1:8545';
const pathToStoreJson = process.argv[3];

encodeBlocks(rpcUrl, pathToStoreJson).catch((reason) => {
  console.error(reason);
  process.exit(1);
});

File: _vendor/gluwa-usc-sdk/package/rust/bin/encode_blocks.rs
use alloy::{
    primitives::B256,
    providers::{Provider, ProviderBuilder, WsConnect},
    rpc::types::{BlockTransactionsKind, TransactionReceipt},
};

use anyhow::Result;
use clap::Parser;
use futures_util::StreamExt;
use hex;

use std::env;
use std::fs;
use std::str::FromStr;
use std::time::SystemTime;

use usc_abi_encoding::abi::abi_encode;
use usc_abi_encoding::common::EncodingVersion;

#[derive(Parser, Debug)]
#[command(name = "encode-blocks")]
pub struct CliArguments {
    #[arg(long, help = "WebSockets URL to an Ethereum RPC", required = true)]
    pub eth_rpc_url: String,

    #[arg(long, help = "Directory path to store JSON files", required = true)]
    pub path_to_store_json: String,
}

async fn encode_transaction(
    provider: impl Provider,
    tx_hash_str: &str,
    rx_or_none: Option<TransactionReceipt>,
) -> Option<String> {
    let tx_hash = B256::from_str(tx_hash_str).unwrap();

    let tx = match provider.get_transaction_by_hash(tx_hash).await {
        Ok(Some(tx)) => tx,
        Ok(None) => {
            println!("    skipping tx {tx_hash_str}: transaction not found via RPC");
            return None;
        }
        Err(e) => {
            println!("    skipping tx {tx_hash_str}: get_transaction_by_hash failed: {e}");
            return None;
        }
    };

    let rx = match rx_or_none {
        Some(rx) => rx,
        None => match provider.get_transaction_receipt(tx_hash).await {
            Ok(Some(rx)) => rx,
            Ok(None) => {
                println!("    skipping tx {tx_hash_str}: receipt not found via RPC");
                return None;
            }
            Err(e) => {
                println!("    skipping tx {tx_hash_str}: get_transaction_receipt failed: {e}");
                return None;
            }
        },
    };

    let encoded_data = abi_encode(tx, rx, EncodingVersion::V1).unwrap();
    let as_str = hex::encode(encoded_data.abi());

    Some(format!("0x{}", as_str))
}

async fn encode_and_write_to_disk(
    path: String,
    provider: impl Provider,
    block_number: u64,
    tx_hash: String,
    rx_or_none: Option<TransactionReceipt>,
) {
    let Some(encoded_data) = encode_transaction(provider, &tx_hash.to_string(), rx_or_none).await
    else {
        // RPC returned None for this transaction (e.g. transient reorg or
        // unindexed tx); skip it rather than writing anything to disk.
        return;
    };

    // <path>/<block_num>/<tx_hash>.txt
    fs::write(
        format!("{}/{}/{}.txt", path, block_number, tx_hash.to_string()),
        encoded_data + "\n",
    )
    .unwrap();
}

async fn block_handler(
    provider: impl Provider + Clone + 'static,
    block_number: u64,
    path_to_store_json: String,
) -> Result<()> {
    println!("new block --- {:?}", block_number);

    fs::create_dir_all(format!("{}/{}", path_to_store_json, block_number))?;

    // Retry the block fetch up to 5 times with a 1s delay; the node may not
    // have the block indexed yet at the moment we get notified.
    const MAX_ATTEMPTS: u32 = 5;
    const RETRY_DELAY: std::time::Duration = std::time::Duration::from_secs(1);
    let mut block = None;
    let mut last_err: Option<anyhow::Error> = None;
    for attempt in 1..=MAX_ATTEMPTS {
        match provider
            .get_block_by_number(block_number.into(), BlockTransactionsKind::Full)
            .await
        {
            Ok(Some(b)) => {
                block = Some(b);
                break;
            }
            Ok(None) => {
                last_err = Some(anyhow::anyhow!(
                    "block {block_number} not found (attempt {attempt}/{MAX_ATTEMPTS})"
                ));
            }
            Err(e) => {
                last_err = Some(anyhow::anyhow!(
                    "failed to fetch block {block_number} (attempt {attempt}/{MAX_ATTEMPTS}): {e}"
                ));
            }
        }
        if attempt < MAX_ATTEMPTS {
            println!("    retrying block --- {block_number} (attempt {attempt}/{MAX_ATTEMPTS})");
            tokio::time::sleep(RETRY_DELAY).await;
        }
    }
    let block = block.ok_or_else(|| {
        last_err.unwrap_or_else(|| anyhow::anyhow!("failed to fetch block {block_number}"))
    })?;

    let tx_hashes = block.transactions.hashes();

    if tx_hashes.len() >= 13 {
        // optimize RPC cost by fetching all receipts at once
        let receipts = provider
            .get_block_receipts(block_number.into())
            .await?
            .unwrap();
        let tasks: Vec<_> = receipts
            .into_iter()
            .map(|rcp| {
                tokio::spawn({
                    let prv = provider.clone();
                    let path = path_to_store_json.clone();
                    async move {
                        encode_and_write_to_disk(
                            path.clone(),
                            prv.clone(),
                            block_number,
                            rcp.transaction_hash.to_string(),
                            Some(rcp),
                        )
                        .await;
                    }
                })
            })
            .collect();

        for task in tasks {
            task.await.unwrap();
        }
    } else {
        // loop over each transaction and fetch its receipt via explicit RPC call
        let tasks: Vec<_> = tx_hashes
            .clone()
            .into_iter()
            .map(|tx_hash| {
                tokio::spawn({
                    let prv = provider.clone();
                    let path = path_to_store_json.clone();
                    async move {
                        encode_and_write_to_disk(
                            path.clone(),
                            prv.clone(),
                            block_number,
                            tx_hash.to_string(),
                            None,
                        )
                        .await;
                    }
                })
            })
            .collect();

        for task in tasks {
            task.await.unwrap();
        }
    }

    println!(
        "<<< done encoding {:?} transactions in block {:?}",
        tx_hashes.len(),
        block_number
    );

    Ok(())
}

#[tokio::main]
async fn main() -> Result<()> {
    let args = CliArguments::parse();

    let start = SystemTime::now();
    let timeout_minutes: u64 = env::var("TIMEOUT_MINUTES").unwrap_or("2".into()).parse()?;
    println!("=== starting with timeout {:?} mins ...", timeout_minutes);

    fs::create_dir_all(args.path_to_store_json.clone())?;

    let provider = ProviderBuilder::new()
        .on_ws(WsConnect::new(args.eth_rpc_url))
        .await?;

    let subscriber = provider.subscribe_blocks().await?;
    let mut stream = subscriber.into_stream();
    let mut last_seen_block: Option<u64> = None;
    while let Some(block) = stream.next().await {
        if last_seen_block == Some(block.number) {
            println!(
                "    skipping duplicate notification for block --- {:?}",
                block.number
            );
            continue;
        }
        last_seen_block = Some(block.number);

        let _ = block_handler(
            provider.clone(),
            block.number,
            args.path_to_store_json.clone(),
        )
        .await;

        let current = start.elapsed()?;
        if current.as_secs() >= timeout_minutes * 60 {
            println!(
                "=== {:?} mins timeout reached. exiting ...",
                timeout_minutes
            );
            break;
        }
    }

    Ok(())
}
